api/types/network/endpoint.go

```diff
@@ -14,7 +14,11 @@ type EndpointSettings struct {
 	IPAMConfig *EndpointIPAMConfig
 	Links      []string
 	Aliases    []string // Aliases holds the list of extra, user-specified DNS names for this endpoint.
-	MacAddress string
+	// DesiredMacAddress is the configured value, it's copied from MacAddress (the
+	// API param field) when the container is created, and cleared in the inspect
+	// output so that only MacAddress is returned to clients. DesiredMacAddress is
+	// used to set the value for netlabel.MacAddress passed to the driver.
+	DesiredMacAddress string `json:",omitempty"`
```

Comment:
I'd even not expose it through the API, otherwise we'll have to support it until the API version v1.44 is dropped:

```suggestion
	DesiredMacAddress string `json:"-"`
```

Reply:
That'd stop it being encoded for persistent storage... so its value wouldn't be preserved over a daemon restart.

So, the idea is to deal with that by including it in the JSON when it has a value - but always clear it in the copy used for 'inspect' responses.

Reply:
Huh, right. I totally forgot we're using the same json tag for both HTTP and storage concerns.

Reply:
Why couldn't this field be added to `EndpointSettings`? 
```go
// EndpointSettings is a package local wrapper for
// networktypes.EndpointSettings which stores Endpoint state that
// needs to be persisted to disk but not exposed in the api.
```

Reply:
I'd forgotten / not-noticed that the comment says those extra fields are persisted - but I don't think they are?

I could be wrong but, I think, the wrapper is always reconstructed here when the container's started ...

https://github.com/moby/moby/blob/794f7127efd841e023dbd88051c577489f6cd3cf/daemon/container_operations.go#L758-L762

Also, the local-wrapper fields aren't available to the daemon's 'inspect' code - so it couldn't differentiate between a generated address and a configured one when it needed to fill in the container-wide `Config.MacAddress`.

(In the first PR, I got the logic wrong for deciding whether the address was configured - it wasn't as similar to the `operIPAM` case as we thought, so `operMAC` wasn't set up properly. And because the field isn't persisted, I don't think it was possible to tell the difference between a generated MAC and a configured one during the container-run, it has to be spotted and remembered during the create.)

If that seems right, I'll fix the comment.

Reply:
> I'd forgotten / not-noticed that the comment says those extra fields are persisted - but I don't think they are?

They're persisted in the container config: `(container.Container).NetworkSettings[].Networks[]`.

> I could be wrong but, I think, the wrapper is always reconstructed here when the container's started ...

So? That could be changed.

> Also, the local-wrapper fields aren't available to the daemon's 'inspect' code - so it couldn't differentiate between a generated address and a configured one when it needed to fill in the container-wide `Config.MacAddress`.

The local-wrapper fields are accessible from the `*Container`. I find it hard to believe that they are not available to the code which maps an ID to a `*Container` and then to an API response.

Reply:
Ok, thank you - that helped me to untangle it a bit.